File: src/klikatech/_1_week/_3_arrays_generics/Example1.scala
package klikatech._1_week._3_arrays_generics

// Arrays
object Example1 {
    // public final int[] arr = new int[]{0, 1, 2, 3, 4}
    val arr: Array[Int] = Array(0, 1, 2, 3, 4)
    val elem2: Int = arr(2) // Java: arr[2]

    def main(args: Array[String]) {
        println("elem2: " + elem2)
    }
}

File: src/klikatech/_7_week/_1_generics/_0_java_arrays/Example1.java
package klikatech._7_week._1_generics._0_java_arrays;

/*
    Java arrays are COVARIANT == type operator T -> T[] are covariant == Child[] <: Parent[]
 */
public class Example1 {

    public static void main(String[] args) {
        String[] strArr = new String[1];
        Object[] objArr = strArr;           // String[] <: Object[]
        objArr[0] = new Integer(42);        // Integer <: Object
        String str = strArr[0];             // Integer(42) => String
        // ArrayStoreException
    }
}
